### PR TITLE
Feature/implement mayo estimator

### PR DESCRIPTION
**Pull request by fmerino21**
### Description
- Basic structure for MAYOEstimator with parameters n (number of variables), m (number of polynomials), o (dimension of the oil space), k (whipping parameter) and q (order of the finite field). 
- Direct attack implementation with time and memory complexities. 
- The direct attack implementation uses `booleansolve_fxl` to find the optimal parameter `K` from the Furue, Nakamura and Takagi strategy. Then, it gets the time and memory complexities from the fastest algorithm to solve an instance of the MQ problem with $m' = m - \frac{kn - K}{m - K} + 1$ equations and $n'  = m' - K$ variables.  

### Review process
- Review the optimization process for the `K` parameter (function `_optimize_k` in `mayo_helper.py`).
- Review the time and memory complexities in the direct attack implementation. 

### Pre-approval checklist
- [ ] The code builds clean without any errors or warnings
- [ ] I've added/updated tests
- [ ] I've added/updated documentation if needed

### CODE REVIEW

File: cryptographic_estimators/MAYOEstimator/MAYOAlgorithms/direct_attack.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    """
+    Construct an instance of DirectAttack estimator
+
+    The most straightforward attack against MAYO is the direct attack, in which the attacker 
+    aims to solve an instance of the MQ problem associated with the public map P^* [FNT21]_.
+
+    INPUT:
+
+    - ``problem`` -- DummyProblem object including all necessary parameters
+    - ``w`` -- linear algebra constant (default: 2)
```
[the remark this answers is not in the document]
Comment (fmerino21) [reply cryptographic_estimators/MAYOEstimator/MAYOAlgorithms/direct_attack.py:38]: In this case the value actually comes from the `MAYOAlgorithm` class, where the value is 2.81 by default, is that ok ?

File: cryptographic_estimators/MAYOEstimator/MAYOAlgorithms/direct_attack.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    - ``h`` -- external hybridization parameter (default: 0)
+    - ``nsolutions`` -- number of solutions in logarithmic scale (default: expected_number_solutions))
+    - ``excluded_algorithms`` -- a list/tuple of MQ algorithms to be excluded (default: [Lokshtanov])
+    - ``memory_access`` -- specifies the memory access cost model (default: 0, choices: 0 - constant, 1 - logarithmic, 2 - square-root, 3 - cube-root or deploy custom function which takes as input the logarithm of the total memory usage)
+    - ``complexity_type`` -- complexity type to consider (0: estimate, 1: tilde O complexity, default: 0)
```
[the remark this answers is not in the document]
Comment (fmerino21) [reply cryptographic_estimators/MAYOEstimator/MAYOAlgorithms/direct_attack.py:43]: As both problems and schemes uses the same method to generate the table and considering that we will implement tilde O complexities in the future, I just force the `complexity_type` parameter to be 0 in the `MAYOAlgorithm` class but the user can still use the parameter show_tilde_o_times=1 when using `.table()` method

**Review by Memphisd — CHANGES_REQUESTED** [review]

File: cryptographic_estimators/MAYOEstimator/mayo_estimator.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+            sage: E.table(show_all_parameters=1)
+            +--------------+-------------+----------------------------+
+            |              |             |          estimate          |
+            +--------------+-------------+------+--------+------------+
+            | algorithm    | attack_type | time | memory | parameters |
+            +--------------+-------------+------+--------+------------+
+            | DirectAttack |   forgery   | 21.4 |   56.9 | {'K': 10}  |
+            +--------------+-------------+------+--------+------------+
+
+            sage: E = MAYOEstimator(n=64, m=60, o=10, k=21, q=16)
```
Comment (Javierverbel) [inline cryptographic_estimators/MAYOEstimator/mayo_estimator.py:92]: Could you add one test per MAYO parameter set? Meaning that one, test for MAYO1, MAYO2, MAYO3 and  MAYO5?

Maybe we can remove some of the long test below.
Please add the tag  `# long time` for test taking more than  5-10 few seconds.

**Review by Javierverbel — CHANGES_REQUESTED** [review]
Please add one test per row in Table5.1: of the MAYO specs

**Review by Javierverbel — DISMISSED** [review]

**Review by Javierverbel — APPROVED** [review]

**Review by Memphisd — APPROVED** [review]
looks good to me

**Pull request merged**